crates/re_query_cache/src/flat_vec_deque.rs

```diff
@@ -0,0 +1,789 @@
+}
+
+// ---
+
+impl<T> FlatVecDeque<T> {
+    /// Prepends an entry comprised of `values` to the deque.
+    ///
+    /// This is the same as `self.insert(0, values)`.
+    ///
+    /// See [`Self::insert`] for more information.
+    #[inline]
+    pub fn push_front(&mut self, values: impl ExactSizeIterator<Item = T>) {
+        self.insert(0, values);
+    }
+
+    /// Appends an entry comprised of `values` to the deque.
+    ///
+    /// This is the same as `self.insert(self.num_entries(), values)`.
+    ///
+    /// See [`Self::insert`] for more information.
+    #[inline]
+    pub fn push_back(&mut self, values: impl ExactSizeIterator<Item = T>) {
+        self.insert(self.num_entries(), values);
+    }
+
+    /// Inserts a single entry at `entry_index`, comprised of the multiple elements given as `values`.
+    ///
+    /// This is O(1) if `entry_index` corresponds to either the start or the end of the deque.
+    /// Otherwise, this requires splitting the deque into two pieces then stitching them back together
+    /// at both ends of the added data.
+    ///
+    /// Panics if `entry_index` is out of bounds.
+    /// Panics if `values` is empty.
+    #[inline]
+    pub fn insert(&mut self, entry_index: usize, values: impl ExactSizeIterator<Item = T>) {
+        let num_values = values.len();
+        let deque = Self {
+            values: values.collect(),
+            offsets: std::iter::once(num_values).collect(),
+        };
+        self.insert_with(entry_index, deque);
+    }
+
+    /// Prepends multiple entries, each comprised of the multiple elements given in `entries`,
+    /// to the deque.
+    ///
+    /// This is the same as `self.insert_range(0, entries)`.
+    ///
+    /// See [`Self::insert_range`] for more information.
+    #[inline]
+    pub fn push_range_front(&mut self, entries: impl IntoIterator<Item = Vec<T>>) {
+        self.insert_range(0, entries);
+    }
+
+    /// Appends multiple entries, each comprised of the multiple elements given in `entries`,
+    /// to the deque.
+    ///
+    /// This is the same as `self.insert_range(self.num_entries(), entries)`.
+    ///
+    /// See [`Self::insert_range`] for more information.
+    #[inline]
+    pub fn push_range_back(&mut self, entries: impl IntoIterator<Item = Vec<T>>) {
+        self.insert_range(self.num_entries(), entries);
+    }
+
+    /// Inserts multiple entries, starting at `entry_index` onwards, each comprised of the multiple elements
+    /// given in `entries`.
+    ///
+    /// This is O(1) if `entry_index` corresponds to either the start or the end of the deque.
+    /// Otherwise, this requires splitting the deque into two pieces then stitching them back together
+    /// at both ends of the added data.
+    ///
+    /// Panics if `entry_index` is out of bounds.
+    /// Panics if any of the value arrays in `entries` is empty.
+    #[inline]
+    pub fn insert_range(&mut self, entry_index: usize, entries: impl IntoIterator<Item = Vec<T>>) {
```

Comment:
The `_range` part of the names is a bit confusing imho. May I suggest `push_many_front`, `push_many_back` and `insert_many` instead?